examples/svelte-kit/src/hooks.server.ts

```diff
@@ -0,0 +1,14 @@
+import type { Handle } from '@sveltejs/kit'
+import cookie from 'cookie'
+import { provideTelefuncContext } from 'telefunc'
+import 'telefunc/async_hooks'
+
+export const handle: Handle = async ({ event, resolve }) => {
+  const cookies = cookie.parse(event.request.headers.get('cookie') || '')
+  event.locals.user = { name: cookies.user ?? 'Elizabeth', id: 0 }
+
+  // Provide telefunc context for all requests
+  provideTelefuncContext(event.locals)
```

Comment:
If we don't use Async Hooks, then we can remove provideTelefuncContext() and instead pass the context to `telefunc({ url, method, body, context })`

Reply:
I see this usage in the doc, but the code does not reference this usage. Am I missing something?

Reply:
Still having some issues:
```
[telefunc@0.1.36][Wrong Usage][getContext()] Make sure to provide a context object, see https://telefunc.com/getContext#provide
Error: [telefunc@0.1.36][Wrong Usage][getContext()] Make sure to provide a context object, see https://telefunc.com/getContext#provide
    at Proxy.getContext (/home/magne/workspace/telefunc/telefunc/dist/cjs/node/server/getContext.js:16:29)
    at Module.onLoadData (/home/magne/workspace/telefunc/examples/svelte-kit/src/routes/Counter.telefunc.ts:11:19)
    at load (/home/magne/workspace/telefunc/examples/svelte-kit/src/routes/+page.ts:6:26)
    at load_data (file:///home/magne/workspace/telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.544_svelte@3.53.1+vite@3.2.3/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:238:38)
    at async file:///home/magne/workspace/telefunc/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.544_svelte@3.53.1+vite@3.2.3/node_modules/@sveltejs/kit/src/runtime/server/page/index.js:170:13
```

You can try latest commit to reproduce the error.

Reply:
It happens during ssr rendering of `src/routes/+page.ts`. `onLoadData` is called, but the `_telefunc` endpoint is not, so context is not set.

Reply:
Yes, it's expected that SSR data isn't supported. (I'll have a look at improving the error message.)

Does SvelteKit have a data fetching hook that is always run on the server-side, like Next.js's `getServerSideProps()`? So that we can use that instead of Telefunc for fetching SSR data.

Reply:
[SvelteKit hooks](https://kit.svelte.dev/docs/hooks) are app-wide only. There is also a possibilility to have a [load hook that is layout-wide](https://kit.svelte.dev/docs/load#cookies-and-headers), but not very useful in this case.
Right now, without calling `provideTelefuncContext`, we wouldn't be able to call a telefunc file in ssr then.

Reply:
By page we can use the `load` function as I did in [the example](https://github.com/brillout/telefunc/pull/54/files#diff-952847c2d8863c2bdae7c43f6a60606ad03907c859936ea7ba2e5bb17ab932b7R5-R10) though.

So yeah:

> We don't need to use Telefunc to fetch the initial data of pages; instead, we can use SvelteKit built-in data fetching mechanism.

I would still personally prefer to have a way to call those files from ssr though, as I do not have to bother to think from where it will be called (and it simplifies refactoring if needed).

Reply:
Defining `load()` in `+page.server.js` instead of `+page.js` enables `load()` to access the database, so that solves the problem.

The problem with `provideTelefuncContext()` (i.e. Async Hooks) is that it doesn't work on the Edge. That's why the default recommendation so far is to use the UI framework instead of Telefunc for SSR data.

RPC + SSR is tricky because of headers (i.e. context) and requires either 1. Async hooks, or 2. deep integration with the UI framework (which needs extensive collaboration with the UI framework).

I plan to create an boilerplate generator `npm init telefunc` which will use the `examples/`. So since `examples/svelte-kit/` will be the official recommendation to get started, I'd rather recommend something that also works on the Edge.

We can keep Async Hooks as long as its maintainability cost stays low. But I'd be inclined to eventually deprecate Async Hooks. I'm open for discussion here and feel free to disagree.

Reply:
> RPC + SSR is tricky because of headers (i.e. context) and requires either 1. Async hooks, or 2. deep integration with the UI framework (which needs extensive collaboration with the UI framework).

We could have integrations with telefunc to fix that on a per framework basis, but I see that could be a lot of work and maintenance for a small benefit.

> I plan to create an boilerplate generator `npm init telefunc` which will use the `examples/`. So since `examples/svelte-kit/` will be the official recommendation to get started, I'd rather recommend something that also works on the Edge.

:+1: 

> We can keep Async Hooks as long as its maintainability cost stays low. But I'd be inclined to eventually deprecate Async Hooks. I'm open for discussion here and feel free to disagree.

I understand. My remark is probably biased because that's how I started using `telefunc`, but if it's clear from the beginning how it should be used, I think it's perfectly fine.

Let's see if I find myself still wanting this in the future, and I'll come up with suggestions then.

Reply:
Sounds good 👍.